app/src/main/scala/org/alephium/explorer/Main.scala

```diff
@@ -55,32 +55,70 @@ class BootUp extends StrictLogging {
   implicit val databaseConfig: DatabaseConfig[PostgresProfile] =
     DatabaseConfig.forConfig[PostgresProfile]("db", typesafeConfig)
 
-  implicit val actorSystem: ActorSystem           = ActorSystem()
-  implicit val executionContext: ExecutionContext = actorSystem.dispatcher
+  def init(): Unit =
+    config.bootMode match {
+      case mode: BootMode.Readable =>
+        initInReadableMode(mode)
 
-  private val explorer: ExplorerState = ExplorerState()
+      case BootMode.WriteOnly =>
+        initInWriteOnlyMode()
+    }
+
+  def initInReadableMode(mode: BootMode.Readable): Unit = {
+    implicit val actorSystem: ActorSystem           = ActorSystem()
+    implicit val executionContext: ExecutionContext = actorSystem.dispatcher
+
+    val explorer: ExplorerState =
+      mode match {
+        case BootMode.ReadOnly  => ExplorerState.ReadOnly()
+        case BootMode.ReadWrite => ExplorerState.ReadWrite()
+      }
 
-  def init(): Unit = {
     explorer
       .start()
       .onComplete {
         case Success(_) => ()
         case Failure(e) =>
-          logger.error(s"Fatal error during initialization: $e")
+          logger.error(s"Fatal error during initialization", e)
           actorSystem.terminate()
       }
 
     Runtime.getRuntime.addShutdownHook(new Thread(() => {
       sideEffect(Await.result(actorSystem.terminate(), 10.seconds))
     }))
+
+    CoordinatedShutdown(actorSystem).addTask(
+      CoordinatedShutdown.PhaseBeforeActorSystemTerminate,
+      "Shutdown services"
+    ) { () =>
+      for {
+        _ <- explorer.stop()
```

Comment:
:+1: That might solve the issue raised by @killerwhile in slack

Reply:
Oh this code is exactly the same as before. No new changes here. Just moved it to a function.

Reply:
ah ok ^^